### PR TITLE
Don't mutate the state object

### PR DESCRIPTION
**Pull request by oampo**
From the [React
docs](https://reactjs.org/docs/react-component.html#state):
> Never mutate this.state directly, as calling setState() afterwards may
replace the mutation you made. Treat this.state as if it were immutable.

### CODE REVIEW

**Comment by tomatau** [comment]
Happy to merge on principle! Also, I like introducing `omit` as it's a good programming term to see in practice even if the implementation is a bit confusing :)

Although, the docs are talking about treating the `this.state` as a mutable property by assignment and skipping the `setState` when you want to change state. 

We don't need to worry about referential equality as long as we use `this.setState()` with the mutated values!

In other words, don't do this:

```js
this.setState({ foo: 'foo', bar: 'bar' })
this.state.foo = 'oof'
this.setState({ bar: 'rab' })
this.state.foo === 'foo' // <-- unintended problem!!
```

This is fine though:

```js
this.setState({ foo: { nestedFoo: 'nested-foo' } })
const { foo } = this.state
delete foo.nestedFoo
this.setState({ foo })
this.state.foo.nestedFoo === undefined // no problems! \o/ 
```

**Comment by tomatau** [comment]
Added you as a collab, feel free to merge

**Comment by oampo** [comment]
Right, I think the subtlety of the rules will probably be lost on students - I'd rather we stick to "don't mutate state on principle".  I'll add a bit on `omit` in the curric text.

**Comment by oampo** [comment]
Can you give me write access so I can merge?

**Comment by tomatau** [comment]
Sure, I think you need to accept the invite first

**Comment by tomatau** [comment]
Btw there are "mutations" happening in the add as well as the map mutates each list object there too

**Comment by tomatau** [comment]
I'll move this to a fork on ThinkfulEd when time permits and update the links inside the curriculum
